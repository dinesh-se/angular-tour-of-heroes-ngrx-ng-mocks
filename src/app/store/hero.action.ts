import { createAction, props } from '@ngrx/store';

import { Hero } from '@model/hero';

export const loadHeroes = createAction('[Heroes] Load Heroes');

export const loadHeroesSuccess = createAction(
  '[Heroes] Load Heroes Success',
  props<{ heroes: Hero[] }>()
);

export const loadHeroesFailure = createAction(
  '[Heroes] Load Heroes Failure',
  props<{ error: any }>()
);

export const searchHeroes = createAction(
  '[Heroes] Search Heroes',
  props<{ query: string }>()
);

export const searchHeroesSuccess = createAction(
  '[Heroes] Search Heroes Success',
  props<{ searchResults: Hero[] }>()
);

export const searchHeroesFailure = createAction(
  '[Heroes] Search Heroes Failure',
  props<{ error: any }>()
);

export const addHero = createAction(
  '[Heroes] Add Hero',
  props<{ name: string }>()
);

export const addHeroSuccess = createAction(
  '[Heroes] Add Hero Success',
  props<{ hero: Hero }>()
);

export const addHeroFailure = createAction(
  '[Heroes] Add Hero Failure',
  props<{ error: any }>()
);

export const updateHero = createAction(
  '[Heroes] Update Hero',
  props<{ hero: Hero }>()
);

export const updateHeroSuccess = createAction(
  '[Heroes] Update Hero Success',
  props<{ hero: Hero }>()
);

export const updateHeroFailure = createAction(
  '[Heroes] Update Hero Failure',
  props<{ error: any }>()
);

export const deleteHero = createAction(
  '[Heroes] Delete Hero',
  props<{ id: number }>()
);

export const deleteHeroSuccess = createAction(
  '[Heroes] Delete Hero Success',
  props<{ id: number }>()
);

export const deleteHeroFailure = createAction(
  '[Heroes] Delete Hero Failure',
  props<{ error: any }>()
);

export const HeroActions = {
  loadHeroes,
  loadHeroesSuccess,
  loadHeroesFailure,
  searchHeroes,
  searchHeroesSuccess,
  searchHeroesFailure,
  addHero,
  addHeroSuccess,
  addHeroFailure,
  updateHero,
  updateHeroSuccess,
  updateHeroFailure,
  deleteHero,
  deleteHeroSuccess,
  deleteHeroFailure
};
