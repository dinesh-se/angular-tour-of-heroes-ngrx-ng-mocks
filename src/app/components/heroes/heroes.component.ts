import { Component, ElementRef, OnInit, ViewChild } from '@angular/core';
import { AsyncPipe, NgFor, NgIf, UpperCasePipe } from '@angular/common';
import { RouterLink } from '@angular/router';

import { Hero } from '@model/hero';
import { HeroDetailComponent } from '@components/hero-detail/hero-detail.component';
import { Store } from '@ngrx/store';
import { HeroesState } from 'src/app/store/hero.reducer';
import { HeroActions } from '@store/hero.action';
import { Observable } from 'rxjs';
import { HeroSelectors } from '@store/hero.selector';

@Component({
  selector: 'app-heroes',
  standalone: true,
  imports: [
    UpperCasePipe,
    NgIf,
    NgFor,
    HeroDetailComponent,
    RouterLink,
    AsyncPipe
  ],
  templateUrl: './heroes.component.html',
  styleUrl: './heroes.component.scss'
})
export class HeroesComponent implements OnInit {
  heroes$!: Observable<Hero[]>;

  @ViewChild('heroName') heroName!: ElementRef;

  constructor(private store: Store<HeroesState>) {}

  ngOnInit(): void {
    this.store.dispatch(HeroActions.loadHeroes());
    this.heroes$ = this.store.select(HeroSelectors.selectAllHeroes);
  }

  add(name: string): void {
    name = name.trim();
    if (!name) { return; }
    this.store.dispatch(HeroActions.addHero({ name }));
    this.heroName.nativeElement.value = '';
  }

  delete(hero: Hero): void {
    this.store.dispatch(HeroActions.deleteHero({ id: hero.id }))
  }
}
