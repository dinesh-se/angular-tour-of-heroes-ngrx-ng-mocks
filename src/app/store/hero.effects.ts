import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { Observable, of } from 'rxjs';
import { catchError, map, switchMap, tap } from 'rxjs/operators';
import { Action } from '@ngrx/store';

import { HeroService } from '@services/hero.service';
import * as HeroActions from './hero.action';

@Injectable()
export class HeroEffects {
  constructor(
    private actions$: Actions,
    private heroService: HeroService
  ) {}

  loadHeroes$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.loadHeroes),
      switchMap(() => this.heroService.getHeroes().pipe(
        tap((heroes) => console.log('HEROES', heroes)),
        map((heroes) => HeroActions.loadHeroesSuccess({ heroes })),
        catchError((error) => of(HeroActions.loadHeroesFailure({ error })))
      ))
    )
  );

  searchHeroes$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.searchHeroes),
      switchMap(({ query }) =>
        this.heroService.searchHeroes(query).pipe(
          map((searchResults) => HeroActions.searchHeroesSuccess({ searchResults })),
          catchError((error) => of(HeroActions.searchHeroesFailure({ error })))
        )
      )
    )
  );

  addHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.addHero),
      switchMap(({ name }) =>
        this.heroService.addHero({ name }).pipe(
          map((addedHero) => HeroActions.addHeroSuccess({ hero: addedHero })),
          catchError((error) => of(HeroActions.addHeroFailure({ error })))
        )
      )
    )
  );

  updateHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.updateHero),
      switchMap(({ hero }) =>
        this.heroService.updateHero(hero).pipe(
          map(() => HeroActions.updateHeroSuccess({ hero })),
          catchError((error) => of(HeroActions.updateHeroFailure({ error })))
        )
      )
    )
  );

  deleteHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.deleteHero),
      switchMap(({ id }) =>
        this.heroService.deleteHero(id).pipe(
          map(() => HeroActions.deleteHeroSuccess({ id })),
          catchError((error) => of(HeroActions.deleteHeroFailure({ error })))
        )
      )
    )
  );
}
