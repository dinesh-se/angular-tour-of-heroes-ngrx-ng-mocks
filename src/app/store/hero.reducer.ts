import { createReducer, on, Action } from '@ngrx/store';

import * as HeroActions from './hero.action';
import { Hero } from '@model/hero';

// Define the shape of the state
export interface HeroesState {
  searchResults: Hero[];
  heroes: Hero[];
  loading: boolean;
  error: any;
}

// Define the initial state
export const initialState: HeroesState = {
  searchResults: [],
  heroes: [],
  loading: false,
  error: null
};

// Create the reducer function using createReducer
const heroesReducer = createReducer(
  initialState,

  // Load Heroes actions
  on(HeroActions.loadHeroes, state => ({
    ...state,
    loading: true,
    error: null
  })),
  on(HeroActions.loadHeroesSuccess, (state, { heroes }) => ({
    ...state,
    heroes,
    loading: false
  })),
  on(HeroActions.loadHeroesFailure, (state, { error }) => ({
    ...state,
    loading: false,
    error
  })),


  on(HeroActions.searchHeroesSuccess, (state, { searchResults }) => ({
    ...state,
    searchResults,
  })),
  on(HeroActions.searchHeroesFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Add Hero actions
  on(HeroActions.addHeroSuccess, (state, { hero }) => ({
    ...state,
    heroes: [...state.heroes, hero]
  })),
  on(HeroActions.addHeroFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Update Hero actions
  on(HeroActions.updateHeroSuccess, (state, { hero }) => ({
    ...state,
    heroes: state.heroes.map(h => (h.id === hero.id ? hero : h))
  })),
  on(HeroActions.updateHeroFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Delete Hero actions
  on(HeroActions.deleteHeroSuccess, (state, { id }) => ({
    ...state,
    heroes: state.heroes.filter(h => h.id !== id)
  })),
  on(HeroActions.deleteHeroFailure, (state, { error }) => ({
    ...state,
    error
  }))
);

// Export the reducer function
export function reducer(state: HeroesState | undefined, action: Action) {
  return heroesReducer(state, action);
}
