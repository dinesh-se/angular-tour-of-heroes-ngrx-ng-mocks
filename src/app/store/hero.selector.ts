import { createSelector, createFeatureSelector } from '@ngrx/store';
import { HeroesState } from './hero.reducer';
import { Hero } from '@model/hero';

export const selectHeroesState = createFeatureSelector<HeroesState>('heroes');

export const selectAllHeroes = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.heroes
);

export const selectHeroesLoading = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.loading
);

export const selectHeroesError = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.error
);

export const selectSearchResults = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.searchResults
);

export const selectQuery = createSelector(
  selectHeroesState,
  (state: HeroesState) =>  state.query
);

export const selectHeroById = (id: number) => createSelector(
  selectAllHeroes,
  (heroes: Hero[]) => heroes.find(hero => hero.id === id)
);
