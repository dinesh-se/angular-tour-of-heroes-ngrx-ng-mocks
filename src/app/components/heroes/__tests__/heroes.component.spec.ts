import { AsyncPipe, NgFor, NgIf, UpperCasePipe } from '@angular/common';
import { Store } from '@ngrx/store';
import { MockStore, createMockStore } from '@ngrx/store/testing';
import { MockBuilder, MockRender, MockedComponentFixture } from 'ng-mocks';

import { HeroesState } from '@store/hero.reducer';
import { HeroSelectors } from '@store/hero.selector';
import { HeroActions } from '@store/hero.action';
import { HeroesComponent } from '../heroes.component';

describe('HeroesComponent', () => {
  let component: HeroesComponent;
  let fixture: MockedComponentFixture<HeroesComponent>;
  const store: MockStore<HeroesState> = createMockStore({
    selectors: [
      {
        selector: HeroSelectors.selectAllHeroes,
        value: [
          {
            id: 1,
            name: 'Hero 1'
          },
          {
            id: 2,
            name: 'Hero 2'
          },
          {
            id: 3,
            name: 'Hero 3'
          }
        ]
      }
    ]
  });
  const dispatchSpy = jest.spyOn(store, 'dispatch');
  const loadHeroesSpy = jest.spyOn(HeroActions, 'loadHeroes');
  const addHeroSpy = jest.spyOn(HeroActions, 'addHero');
  const DeleteHeroSpy = jest.spyOn(HeroActions, 'deleteHero');

  beforeEach(async () => {
    await MockBuilder(HeroesComponent)
      .mock(Store, store)
      .keep(NgIf)
      .keep(NgFor)
      .keep(UpperCasePipe)
      .keep(AsyncPipe);

    fixture = MockRender(HeroesComponent);
    component = fixture.point.componentInstance;
    fixture.detectChanges();
  });

  afterEach(() => {
    jest.clearAllMocks();
  })

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(dispatchSpy).toHaveBeenCalled();
    expect(loadHeroesSpy).toHaveBeenCalled();
    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch addHero action on clicking add button', () => {
    const el = fixture.point.nativeElement;
    const heroNameInput = el.querySelector('#new-hero');
    const addBtn = el.querySelector('.add-btn');

    heroNameInput.value = 'New Hero ';
    addBtn.click();

    expect(dispatchSpy).toHaveBeenCalledTimes(2);
    expect(addHeroSpy).toHaveBeenCalledWith({ name: 'New Hero' });

    expect(component.heroName.nativeElement.value).toBe('');
  });

  it('should not add Hero when input is empty', () => {
    const el = fixture.point.nativeElement;
    const heroNameInput = el.querySelector('#new-hero');
    const addBtn = el.querySelector('.add-btn');

    heroNameInput.value = '';
    addBtn.click();

    expect(dispatchSpy).toHaveBeenCalledTimes(1);
    expect(addHeroSpy).not.toHaveBeenCalled();
  });

  it('should dispatch deleteHero action on clicking delete button', () => {
    const el = fixture.point.nativeElement;
    const firstDeleteBtn = el.querySelectorAll('.delete-btn')[0];

    firstDeleteBtn.click();

    expect(dispatchSpy).toHaveBeenCalledTimes(2);
    expect(DeleteHeroSpy).toHaveBeenCalledWith({ id: 1 });
  });
});
