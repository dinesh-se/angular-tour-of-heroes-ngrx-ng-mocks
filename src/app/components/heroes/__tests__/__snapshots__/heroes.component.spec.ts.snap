// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroesComponent should create 1`] = `
<mock-render>
  <app-heroes>
    <h2>
      My Heroes
    </h2>
    <div>
      <label
        for="new-hero"
      >
        Hero name: 
      </label>
      <input
        id="new-hero"
      />
      <button
        class="add-btn"
        type="button"
      >
         Add hero 
      </button>
    </div>
    <ul
      class="heroes"
    >
      <li>
        <a>
          <span
            class="badge"
          >
            1
          </span>
           Hero 1 
        </a>
        <button
          class="delete-btn"
          title="delete hero"
          type="button"
        >
          x
        </button>
      </li>
      <li>
        <a>
          <span
            class="badge"
          >
            2
          </span>
           Hero 2 
        </a>
        <button
          class="delete-btn"
          title="delete hero"
          type="button"
        >
          x
        </button>
      </li>
      <li>
        <a>
          <span
            class="badge"
          >
            3
          </span>
           Hero 3 
        </a>
        <button
          class="delete-btn"
          title="delete hero"
          type="button"
        >
          x
        </button>
      </li>
    </ul>
  </app-heroes>
</mock-render>
`;
